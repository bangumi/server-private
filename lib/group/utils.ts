import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
