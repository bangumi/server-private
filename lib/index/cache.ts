export function getSlimCacheKey(id: number): string {
  return `idx:v4:slim:${id}`;
}
