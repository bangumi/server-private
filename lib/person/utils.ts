import { db, op, schema } from '@app/drizzle';

import type { PersonCat } from './type';

export async function getPersonCollect(
  cat: PersonCat,
  userID: number,
  personID: number,
): Promise<number | null> {
  const [d] = await db
    .select()
    .from(schema.chiiPersonCollects)
    .where(
      op.and(
        op.eq(schema.chiiPersonCollects.cat, cat),
        op.eq(schema.chiiPersonCollects.uid, userID),
        op.eq(schema.chiiPersonCollects.mid, personID),
      ),
    )
    .limit(1);
  if (!d) {
    return null;
  }
  return d.createdAt;
}
