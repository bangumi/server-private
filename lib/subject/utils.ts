import * as php from '@trim21/php-serialize';
import { DateTime } from 'luxon';

import { db, decr, incr, op, type orm, schema, type Txn } from '@app/drizzle';

import { type CollectionType, SubjectType, type UserEpisodeStatusItem } from './type';
import { EpisodeCollectionStatus, getCollectionTypeField } from './type';

/** 更新条目收藏计数，需要在事务中执行 */
export async function updateSubjectCollection(
  t: Txn,
  subjectID: number,
  newType: CollectionType,
  oldType?: CollectionType,
) {
  if (oldType && oldType === newType) {
    return;
  }
  const toUpdate: Record<string, op.SQL> = {};
  toUpdate[getCollectionTypeField(newType)] = incr(
    schema.chiiSubjects[getCollectionTypeField(newType)],
  );
  if (oldType) {
    toUpdate[getCollectionTypeField(oldType)] = decr(
      schema.chiiSubjects[getCollectionTypeField(oldType)],
    );
  }
  await t
    .update(schema.chiiSubjects)
    .set(toUpdate)
    .where(op.eq(schema.chiiSubjects.id, subjectID))
    .limit(1);
}

function getRatingField(rate: number) {
  return [
    undefined,
    'rate1',
    'rate2',
    'rate3',
    'rate4',
    'rate5',
    'rate6',
    'rate7',
    'rate8',
    'rate9',
    'rate10',
  ][rate];
}

/** 更新条目评分，需要在事务中执行 */
export async function updateSubjectRating(
  t: Txn,
  subjectID: number,
  oldRate: number,
  newRate: number,
) {
  if (oldRate === newRate) {
    return;
  }
  const newField = getRatingField(newRate);
  const oldField = getRatingField(oldRate);
  const toUpdate: Record<string, op.SQL> = {};
  if (newField) {
    const field = newField as keyof orm.ISubjectFields;
    toUpdate[field] = incr(schema.chiiSubjectFields[field]);
  }
  if (oldField) {
    const field = oldField as keyof orm.ISubjectFields;
    toUpdate[field] = decr(schema.chiiSubjectFields[field]);
  }
  if (Object.keys(toUpdate).length === 0) {
    return;
  }
  await t
    .update(schema.chiiSubjectFields)
    .set(toUpdate)
    .where(op.eq(schema.chiiSubjectFields.id, subjectID))
    .limit(1);
}

/** 更新条目剧集进度，需要在事务中执行 */
export async function updateSubjectEpisodeProgress(
  t: Txn,
  userID: number,
  subjectID: number,
  episodeID: number,
  type: number,
): Promise<number> {
  const [current] = await t
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    );
  let watchedEpisodes = 0;
  if (current?.status) {
    const epStatusList = parseSubjectEpStatus(current.status);
    epStatusList[episodeID] = {
      eid: episodeID,
      type,
    };
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t
      .update(schema.chiiEpStatus)
      .set({ status: newStatus, updatedAt: DateTime.now().toUnixInteger() })
      .where(op.eq(schema.chiiEpStatus.id, current.id))
      .limit(1);
  } else {
    const epStatusList: Record<number, UserEpisodeStatusItem> = {};
    epStatusList[episodeID] = {
      eid: episodeID,
      type,
    };
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t.insert(schema.chiiEpStatus).values({
      uid: userID,
      sid: subjectID,
      status: newStatus,
      updatedAt: DateTime.now().toUnixInteger(),
    });
  }
  return watchedEpisodes;
}

/** 标记条目剧集为已观看，需要在事务中执行 */
export async function markEpisodesAsWatched(
  t: Txn,
  userID: number,
  subjectID: number,
  episodeIDs: number[],
  revertOthers = false,
): Promise<number> {
  const epStatusList: Record<number, UserEpisodeStatusItem> = {};
  for (const episodeID of episodeIDs) {
    epStatusList[episodeID] = {
      eid: episodeID,
      type: EpisodeCollectionStatus.Done,
    };
  }
  const [current] = await t
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    );
  let watchedEpisodes = 0;
  if (current?.status) {
    const oldList = parseSubjectEpStatus(current.status);
    for (const x of Object.values(oldList)) {
      if (episodeIDs.includes(x.eid)) {
        continue;
      }
      if (revertOthers && x.type === EpisodeCollectionStatus.Done) {
        epStatusList[x.eid] = {
          eid: x.eid,
          type: EpisodeCollectionStatus.None,
        };
      } else {
        epStatusList[x.eid] = x;
      }
    }
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t
      .update(schema.chiiEpStatus)
      .set({ status: newStatus, updatedAt: DateTime.now().toUnixInteger() })
      .where(op.eq(schema.chiiEpStatus.id, current.id))
      .limit(1);
  } else {
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t.insert(schema.chiiEpStatus).values({
      uid: userID,
      sid: subjectID,
      status: newStatus,
      updatedAt: DateTime.now().toUnixInteger(),
    });
  }
  return watchedEpisodes;
}

export function parseSubjectEpStatus(status: string): Record<number, UserEpisodeStatusItem> {
  const result: Record<number, UserEpisodeStatusItem> = {};
  if (!status) {
    return result;
  }
  const epStatusList = php.parse(status) as Record<number, UserEpisodeStatusItem>;
  for (const x of Object.values(epStatusList)) {
    result[x.eid] = x;
  }
  return result;
}

export async function getEpStatus(
  userID: number,
  subjectID: number,
): Promise<Record<number, UserEpisodeStatusItem>> {
  const [data] = await db
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    )
    .limit(1);
  if (!data) {
    return {};
  }
  return parseSubjectEpStatus(data.status);
}

/** 完成条目进度，需要在事务中执行 */
export async function completeSubjectProgress(
  t: Txn,
  userID: number,
  subject: orm.ISubject,
  interest: Partial<orm.ISubjectInterest>,
) {
  if (subject.eps > 0) {
    interest.epStatus = subject.eps;
  }
  if (subject.volumes > 0) {
    interest.volStatus = subject.volumes;
  }
  if ([SubjectType.Anime, SubjectType.Real].includes(subject.typeID)) {
    const episodes = await t
      .select({ id: schema.chiiEpisodes.id })
      .from(schema.chiiEpisodes)
      .where(
        op.and(
          op.eq(schema.chiiEpisodes.subjectID, subject.id),
          op.eq(schema.chiiEpisodes.type, 0),
          op.eq(schema.chiiEpisodes.ban, 0),
        ),
      );
    const episodeIDs = episodes.map((e) => e.id);
    if (episodeIDs.length > 0) {
      await markEpisodesAsWatched(t, userID, subject.id, episodeIDs);
    }
  }
}
