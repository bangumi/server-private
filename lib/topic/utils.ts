import type * as orm from '@app/drizzle/orm.ts';
import { HotGroups } from '@app/lib/group/consts.ts';

export function scoredUpdateTime(timestamp: number, topic: orm.IGroupTopic): number {
  if (HotGroups.includes(topic.gid) && topic.replies > 0) {
    const createdAt = topic.createdAt;
    const createdHours = (timestamp - createdAt) / 3600;
    const gravity = 1.8;
    const baseScore = (Math.pow(createdHours + 0.1, gravity) / topic.replies) * 200;
    const scoredLastPost = Math.trunc(timestamp - baseScore);
    return Math.min(scoredLastPost, timestamp);
  }

  return timestamp;
}
