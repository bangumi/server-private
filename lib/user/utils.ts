import { db, op, schema } from '@app/drizzle';
import { UnexpectedNotFoundError } from '@app/lib/error.ts';
import redis from '@app/lib/redis.ts';
import * as fetcher from '@app/lib/types/fetcher.ts';
import type * as res from '@app/lib/types/res.ts';
import {
  getFollowersCacheKey,
  getFriendsCacheKey,
  getRelationCacheKey,
} from '@app/lib/user/cache.ts';

/** Cached: Get friend ids of user(uid) */
export async function fetchFriends(uid?: number): Promise<number[]> {
  if (!uid) {
    return [];
  }

  const cached = await redis.get(getFriendsCacheKey(uid));
  if (cached) {
    return JSON.parse(cached) as number[];
  }

  const friends = await db
    .select({ fid: schema.chiiFriends.fid })
    .from(schema.chiiFriends)
    .where(op.eq(schema.chiiFriends.uid, uid));
  const result = friends.map((x) => x.fid);
  await redis.setex(getFriendsCacheKey(uid), 3600, JSON.stringify(result));
  return result;
}

/** Cached: Get follower ids of user(uid) */
export async function fetchFollowers(uid?: number): Promise<number[]> {
  if (!uid) {
    return [];
  }

  const cached = await redis.get(getFollowersCacheKey(uid));
  if (cached) {
    return JSON.parse(cached) as number[];
  }

  const followers = await db
    .select({ uid: schema.chiiFriends.uid })
    .from(schema.chiiFriends)
    .where(op.eq(schema.chiiFriends.fid, uid));
  const result = followers.map((x) => x.uid);
  await redis.setex(getFollowersCacheKey(uid), 3600, JSON.stringify(result));
  return result;
}

/** Cached: Is user(another) is friend of user(uid) */
export async function isFriends(uid: number, another: number): Promise<boolean> {
  const cached = await redis.get(getRelationCacheKey(uid, another));
  if (cached) {
    return cached === '1';
  }

  const [d] = await db
    .select({ uid: schema.chiiFriends.uid, fid: schema.chiiFriends.fid })
    .from(schema.chiiFriends)
    .where(op.and(op.eq(schema.chiiFriends.uid, uid), op.eq(schema.chiiFriends.fid, another)))
    .limit(1);
  const result = d ? 1 : 0;
  await redis.setex(getRelationCacheKey(uid, another), 3600, result);
  return result === 1;
}

export function ghostUser(uid: number): res.ISlimUser {
  return {
    id: 0,
    username: uid.toString(),
    nickname: `deleted or missing user ${uid}`,
    avatar: {
      small: '',
      medium: '',
      large: '',
    },
    group: 0,
    sign: '',
    joinedAt: 0,
  };
}

/** Cached */
export async function fetchUserX(uid: number): Promise<res.ISlimUser> {
  const user = await fetcher.fetchSlimUserByID(uid);
  if (!user) {
    throw new UnexpectedNotFoundError(`user ${uid} not found`);
  }
  return user;
}
