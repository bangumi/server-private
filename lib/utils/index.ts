import * as crypto from 'node:crypto';

import { customAlphabet } from 'nanoid/async';

const base62Chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
if (base62Chars.length !== 62) {
  throw new TypeError('characters is not 62 length');
}

const generator = customAlphabet(base62Chars, 32);

export const randomBase62String = (size: number) => generator(size);

export async function randomBytes(size: number): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    crypto.randomBytes(size, (err, buf) => {
      if (err) {
        reject(err);
      } else {
        resolve(buf);
      }
    });
  });
}

/**
 * Parse string as int, strictly
 *
 * 避免出现 `parseInt('1abc') === 1` 的问题
 */
export function intval(value: string | number): number {
  const v = parseIntStrict(value);
  if (v === undefined) {
    throw new Error(`"${value}" is not a valid integer`);
  }
  return v;
}

function parseIntStrict(integer: string | number) {
  if (typeof integer === 'number') {
    return Number.isInteger(integer) ? integer : undefined;
  }

  const n = Number(integer);

  if (Number.isInteger(n)) {
    return n;
  }
}

export function md5(s: string) {
  return crypto.createHash('md5').update(s).digest('hex');
}
