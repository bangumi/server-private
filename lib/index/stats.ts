import { DateTime } from 'luxon';

import { db, op, schema } from '@app/drizzle';
import { IndexRelatedCategory } from '@app/lib/index/types.ts';
import { SubjectType } from '@app/lib/subject/type.ts';
import type { IIndexStats } from '@app/lib/types/res';
import { decode } from '@app/lib/utils';

interface IInnerIndexStats {
  '1'?: number;
  '2'?: number;
  '3'?: number;
  '4'?: number;
  '6'?: number;
  character?: number;
  person?: number;
  ep?: number;
  blog?: number;
  group_topic?: number;
  subject_topic?: number;
}

export async function updateIndexStats(indexId: number) {
  const now = DateTime.now().toUnixInteger();
  await db.transaction(async (tx) => {
    const data = await tx
      .select({
        cat: schema.chiiIndexRelated.cat,
        type: schema.chiiIndexRelated.type,
        count: op.countDistinct(schema.chiiIndexRelated.sid),
      })
      .from(schema.chiiIndexRelated)
      .where(
        op.and(op.eq(schema.chiiIndexRelated.rid, indexId), op.eq(schema.chiiIndexRelated.ban, 0)),
      )
      .groupBy(schema.chiiIndexRelated.cat, schema.chiiIndexRelated.type);

    const stats: IInnerIndexStats = {};
    let total = 0;

    for (const { cat, type, count } of data) {
      total += count;
      if (cat === IndexRelatedCategory.Subject) {
        switch (type) {
          case SubjectType.Book: {
            stats['1'] = count;
            break;
          }
          case SubjectType.Anime: {
            stats['2'] = count;
            break;
          }
          case SubjectType.Music: {
            stats['3'] = count;
            break;
          }
          case SubjectType.Game: {
            stats['4'] = count;
            break;
          }
          case SubjectType.Real: {
            stats['6'] = count;
            break;
          }
        }
      } else {
        switch (cat) {
          case IndexRelatedCategory.Character: {
            stats.character = count;
            break;
          }
          case IndexRelatedCategory.Person: {
            stats.person = count;
            break;
          }
          case IndexRelatedCategory.Ep: {
            stats.ep = count;
            break;
          }
          case IndexRelatedCategory.Blog: {
            stats.blog = count;
            break;
          }
          case IndexRelatedCategory.GroupTopic: {
            stats.group_topic = count;
            break;
          }
          case IndexRelatedCategory.SubjectTopic: {
            stats.subject_topic = count;
            break;
          }
        }
      }
    }

    const statsString = JSON.stringify(stats);

    await tx
      .update(schema.chiiIndexes)
      .set({ stats: statsString, total, updatedAt: now })
      .where(op.eq(schema.chiiIndexes.id, indexId));
  });
}

export function parseIndexStats(stats: string): IIndexStats {
  const innerStats = decode(stats) as IInnerIndexStats;
  return {
    subject: {
      book: innerStats['1'],
      anime: innerStats['2'],
      music: innerStats['3'],
      game: innerStats['4'],
      real: innerStats['6'],
    },
    character: innerStats.character,
    person: innerStats.person,
    episode: innerStats.ep,
    blog: innerStats.blog,
    groupTopic: innerStats.group_topic,
    subjectTopic: innerStats.subject_topic,
  };
}
