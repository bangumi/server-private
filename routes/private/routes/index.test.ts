import { describe, expect, test } from 'vitest';

import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './index.ts';

describe('get index', () => {
  test('should get index', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/indexes/15045',
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get index related', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/indexes/15045/related',
    });
    expect(res.json()).toMatchSnapshot();
  });
});
