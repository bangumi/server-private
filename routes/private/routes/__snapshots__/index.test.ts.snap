// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get index > should get index 1`] = `
Object {
  "award": 0,
  "collects": 96,
  "createdAt": 1352366596,
  "desc": "[url]http://www.tudou.com/programs/view/W6eIoxnHs6g/[/url]
有美国动画混入，所以准确的说是在日本播放的动画最高收视率（而且是关东地区的
基本大部分是70年代的，那个年代娱乐贫乏优势真大",
  "id": 15045,
  "replies": 8,
  "stats": Object {
    "2": 101,
  },
  "title": "日本动画最高收视率TOP100",
  "total": 101,
  "type": 0,
  "uid": 14127,
  "updatedAt": 1356922367,
  "user": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
      "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
      "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
    },
    "group": 0,
    "id": 14127,
    "joinedAt": 0,
    "nickname": "nickname 14127",
    "sign": "sing 14127",
    "username": "14127",
  },
}
`;

exports[`get index > should get index related 1`] = `
Object {
  "data": Array [
    Object {
      "award": "",
      "cat": 0,
      "comment": "1位 40.3",
      "createdAt": 1352382566,
      "id": 29277,
      "order": 1,
      "rid": 15045,
      "sid": 806,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "2位 39.9",
      "createdAt": 1352382511,
      "id": 29276,
      "order": 2,
      "rid": 15045,
      "sid": 28165,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "3位 39.4",
      "createdAt": 1352382444,
      "id": 29275,
      "order": 3,
      "rid": 15045,
      "sid": 32585,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "4位 36.9",
      "createdAt": 1352382378,
      "id": 29274,
      "order": 4,
      "rid": 15045,
      "sid": 47243,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "5位 36.7",
      "createdAt": 1352382135,
      "id": 29272,
      "order": 5,
      "rid": 15045,
      "sid": 41983,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "5位 36.7",
      "createdAt": 1352382314,
      "id": 29273,
      "order": 5,
      "rid": 15045,
      "sid": 53805,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "7位 36",
      "createdAt": 1352382072,
      "id": 29271,
      "order": 7,
      "rid": 15045,
      "sid": 37258,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "8位 35.6",
      "createdAt": 1352382022,
      "id": 29270,
      "order": 8,
      "rid": 15045,
      "sid": 53804,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "9位 34.5",
      "createdAt": 1352381764,
      "id": 29266,
      "order": 9,
      "rid": 15045,
      "sid": 53739,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "10位 33.7 （有奇怪的美国动画混进去了！",
      "createdAt": 1352381582,
      "id": 29264,
      "order": 10,
      "rid": 15045,
      "sid": 25713,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "11位 33.6",
      "createdAt": 1352380988,
      "id": 29258,
      "order": 11,
      "rid": 15045,
      "sid": 53802,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "12位 32.5",
      "createdAt": 1352380796,
      "id": 29257,
      "order": 12,
      "rid": 15045,
      "sid": 36359,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "13位 31.9",
      "createdAt": 1352380648,
      "id": 29252,
      "order": 13,
      "rid": 15045,
      "sid": 53801,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "13位 31.9",
      "createdAt": 1352380716,
      "id": 29254,
      "order": 13,
      "rid": 15045,
      "sid": 2741,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "15位 31.8",
      "createdAt": 1352380464,
      "id": 29246,
      "order": 15,
      "rid": 15045,
      "sid": 53800,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "16位 31.6",
      "createdAt": 1352380102,
      "id": 29245,
      "order": 16,
      "rid": 15045,
      "sid": 10404,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "17位 31.2",
      "createdAt": 1352379920,
      "id": 29241,
      "order": 17,
      "rid": 15045,
      "sid": 37460,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "18位 30.7",
      "createdAt": 1352379781,
      "id": 29239,
      "order": 18,
      "rid": 15045,
      "sid": 53794,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "19位 30.4",
      "createdAt": 1352379633,
      "id": 29235,
      "order": 19,
      "rid": 15045,
      "sid": 10390,
      "type": 2,
    },
    Object {
      "award": "",
      "cat": 0,
      "comment": "20位 30.1",
      "createdAt": 1352379532,
      "id": 29234,
      "order": 20,
      "rid": 15045,
      "sid": 10374,
      "type": 2,
    },
  ],
  "total": 101,
}
`;
